File: code/packages/go/in-memory-data-store-engine/engine.go
package datastoreengine

import (
	"strings"
	"sync"
	"sync/atomic"

	datastoreprotocol "github.com/adhithyan15/coding-adventures/code/packages/go/in-memory-data-store-protocol"
	resp "github.com/adhithyan15/coding-adventures/code/packages/go/resp-protocol"
)

type CommandHandler func(store *Store, args [][]byte) (*Store, resp.Value)

type CommandRegistration struct {
	Handler        CommandHandler
	Mutating       bool
	SkipLazyExpire bool
}

type DataStoreBackend interface {
	ExecuteFrame(frame datastoreprotocol.CommandFrame) resp.Value
	ExecuteParts(parts [][]byte) resp.Value
	Store() *Store
	ActiveExpireAll()
}

type DataStoreEngine struct {
	mu       sync.Mutex
	store    *Store
	commands map[string]CommandRegistration
	frozen   atomic.Bool
}

func New() *DataStoreEngine {
	engine := &DataStoreEngine{
		store:    NewStore(),
		commands: make(map[string]CommandRegistration),
	}
	registerBuiltinCommands(engine)
	return engine
}

func FromStore(store *Store) *DataStoreEngine {
	engine := &DataStoreEngine{
		store:    store.Clone(),
		commands: make(map[string]CommandRegistration),
	}
	registerBuiltinCommands(engine)
	return engine
}

func (e *DataStoreEngine) RegisterCommand(name string, mutating bool, skipLazyExpire bool, handler CommandHandler) {
	if e.frozen.Load() {
		panic("cannot register commands on a frozen engine")
	}
	e.mu.Lock()
	defer e.mu.Unlock()
	e.commands[strings.ToUpper(name)] = CommandRegistration{
		Handler:        handler,
		Mutating:       mutating,
		SkipLazyExpire: skipLazyExpire,
	}
}

func (e *DataStoreEngine) Freeze() {
	e.frozen.Store(true)
}

func (e *DataStoreEngine) IsFrozen() bool {
	return e.frozen.Load()
}

func (e *DataStoreEngine) ExecuteFrame(frame datastoreprotocol.CommandFrame) resp.Value {
	_, response := e.executeWithDB(e.currentDB(), frame.Command, frame.Args, true)
	return response
}

func (e *DataStoreEngine) ExecuteParts(parts [][]byte) resp.Value {
	frame, ok := datastoreprotocol.FromParts(parts)
	if !ok {
		return resp.ErrorValue("ERR empty command")
	}
	return e.ExecuteFrame(frame)
}

func (e *DataStoreEngine) Store() *Store {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.store.Clone()
}

func (e *DataStoreEngine) ActiveExpireAll() {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.store = e.store.ActiveExpireAll()
}

func (e *DataStoreEngine) executeWithDB(dbIndex int, command string, args [][]byte, recordAOF bool) (int, resp.Value) {
	e.mu.Lock()
	registration, ok := e.commands[strings.ToUpper(command)]
	if !ok {
		e.mu.Unlock()
		return dbIndex, resp.ErrorValue("ERR unknown command '" + strings.ToUpper(command) + "'")
	}

	store := e.store.Clone().WithActiveDB(dbIndex)
	e.mu.Unlock()

	if !registration.SkipLazyExpire && len(args) > 0 {
		store = store.ExpireLazy(args[0])
	}

	newStore, response := registration.Handler(store, args)
	if newStore == nil {
		newStore = store
	}

	e.mu.Lock()
	e.store = newStore
	activeDB := newStore.ActiveDB
	e.mu.Unlock()

	return activeDB, response
}

func (e *DataStoreEngine) currentDB() int {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.store.ActiveDB
}

File: code/packages/typescript/in-memory-data-store-engine/src/engine.ts
import { type DataStoreCommand, commandName } from "@coding-adventures/in-memory-data-store-protocol";
import { type RespValue, errorValue } from "@coding-adventures/resp-protocol";
import { Store } from "./store.js";
import { installDefaultCommands, isMutatingCommand, type CommandHandler } from "./commands.js";

const LAZY_EXPIRE_EXEMPT = new Set([
  "PING",
  "ECHO",
  "SELECT",
  "INFO",
  "DBSIZE",
  "FLUSHDB",
  "FLUSHALL",
  "KEYS",
]);

export interface DataStoreModule {
  register(engine: DataStoreEngine): void;
}

export class DataStoreEngine {
  private storeState: Store;
  private readonly handlers = new Map<string, CommandHandler>();

  constructor(store: Store = Store.empty()) {
    this.storeState = store;
    installDefaultCommands((name, handler) => {
      this.registerCommand(name, handler);
    });
  }

  get store(): Store {
    return this.storeState;
  }

  registerCommand(name: string, handler: CommandHandler): this {
    this.handlers.set(name.toUpperCase(), handler);
    return this;
  }

  registerModule(module: DataStoreModule): this {
    module.register(this);
    return this;
  }

  execute(command: DataStoreCommand | string[]): RespValue {
    const parts = Array.isArray(command) ? command : [command.name, ...command.args];
    if (parts.length === 0) {
      return errorValue("ERR empty command");
    }
    const name = commandName(parts);
    const handler = this.handlers.get(name);
    if (!handler) {
      return errorValue(`ERR unknown command '${name}'`);
    }

    const shouldLazyExpire = !LAZY_EXPIRE_EXEMPT.has(name) && parts.length > 1;
    const inputStore = shouldLazyExpire ? this.storeState.expireLazy(parts[1]) : this.storeState;
    const [nextStore, response] = handler(inputStore, parts.slice(1));
    this.storeState = nextStore;
    return response;
  }

  executeCommand(command: DataStoreCommand): RespValue {
    return this.execute(command);
  }

  executeParts(parts: string[]): RespValue {
    return this.execute(parts);
  }

  reset(store: Store): void {
    this.storeState = store;
  }
}

export function createDataStoreEngine(store: Store = Store.empty()): DataStoreEngine {
  return new DataStoreEngine(store);
}

export { isMutatingCommand };
